helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -0,0 +1,247 @@
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+    if (clusterChanges.containsKey(ClusterDataDetector.ChangeType.ClusterConfigChange)
+        || clusterChanges.containsKey(ClusterDataDetector.ChangeType.InstanceConfigChange)
+        || clusterChanges.containsKey(ClusterDataDetector.ChangeType.BaselineAssignmentChange)) {
+      // If the cluster topology or baseline assignment has been modified, need to reassign all replicas
+      toBeAssignedReplicas
+          .addAll(replicaMap.values().stream().flatMap(Set::stream).collect(Collectors.toSet()));
+    } else {
+      // check each resource to identify the allocated replicas and to-be-assigned replicas.
+      for (String resourceName : replicaMap.keySet()) {
+        Set<AssignableReplica> replicas = replicaMap.get(resourceName);
+        // 1. if the resource config/idealstate is changed, need to reassign.
+        // 2. if the resource does appear in the best possible assignment, need to reassign.
+        if (clusterChanges.getOrDefault(ClusterDataDetector.ChangeType.ResourceConfigChange,
+            Collections.emptySet()).contains(resourceName) || clusterChanges
+            .getOrDefault(ClusterDataDetector.ChangeType.ResourceIdealStatesChange,
+                Collections.emptySet()).contains(resourceName) || !bestPossibleAssignment
+            .containsKey(resourceName)) {
+          toBeAssignedReplicas.addAll(replicas);
+          continue; // go to check next resource
+        } else {
+          // check for every best possible assignments to identify if the related replicas need to reassign.
+          ResourceAssignment assignment = bestPossibleAssignment.get(resourceName);
+          // <partition, <instance, state>>
+          Map<String, Map<String, String>> stateMap = assignment.getMappedPartitions().stream()
+              .collect(Collectors.toMap(partition -> partition.getPartitionName(),
+                  partition -> new HashMap<>(assignment.getReplicaMap(partition))));
+          for (AssignableReplica replica : replicas) {
+            // Find any ACTIVE instance allocation that has the same state with the replica
+            Optional<Map.Entry<String, String>> instanceNameOptional =
+                stateMap.getOrDefault(replica.getPartitionName(), Collections.emptyMap()).entrySet()
+                    .stream().filter(instanceStateMap ->
+                    instanceStateMap.getValue().equals(replica.getReplicaState()) && activeInstances
+                        .contains(instanceStateMap.getKey())).findAny();
+            // 3. if no such an instance in the bestPossible assignment, need to reassign the replica
+            if (!instanceNameOptional.isPresent()) {
+              toBeAssignedReplicas.add(replica);
+              continue; // go to check the next replica
+            } else {
+              String instanceName = instanceNameOptional.get().getKey();
```

Comment:
Just a remind here. You may get NoSuchElementException when it is empty. Be ready to handle the exception outside, otherwise you may break the pipeline.

Or you can use Optional.isPresent() to check and handle it at this layer.

Reply:
That has been checked in the "if" condition already.